pkg/controller/architecture/architecture.go

```diff
@@ -0,0 +1,92 @@
+package architecture
+
+import (
+	"fmt"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+var _ NodeArchitectureMap = &defaultNodeMap{}
+
+const (
+	invalidFuncInput = "invalid function input"
+)
+
+// NodeMetadata metadata about a particular node
+type NodeMetadata struct {
+	OS           string
+	Architecture string
+}
+
+type NodeArchitectureMap interface {
+	GetNodeArchitecture(node string) (*NodeMetadata, error)
+	AddNode(node *corev1.Node) error
+	DeleteNode(node string) error
+	Length() int
+}
+
+type defaultNodeMap struct {
+	mu    sync.RWMutex
+	nodes map[string]*NodeMetadata
+}
+
+func New() *defaultNodeMap {
+	// might need to pass an initial map
+	return &defaultNodeMap{
+		nodes: make(map[string]*NodeMetadata),
+	}
+}
+
+func (m *defaultNodeMap) GetNodeArchitecture(node string) (*NodeMetadata, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if _, ok := m.nodes[node]; !ok {
+		// no data about the node was found, return error
+		return nil, fmt.Errorf("error fetching node's architecture data")
```

Comment:
Not a huge fan of returning an error here, since it's not so much an error, but it not existing in the map.

Better to return a bool instead, and the consumer can choose whether to error or not.